### PR TITLE
json: update json doc

### PR DESCRIPTION
**Pull request by xiongjiwei**
### First-time contributors' checklist 

- [x] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v6.3 (TiDB 6.3 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)

### What is the related PR or file link(s)?

- This PR is translated from:
- Other reference link(s):

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

File: data-type-json.md [hunk 1]
```diff
@@ -29,4 +25,85 @@ INSERT INTO city (id,detail) VALUES (1, '{"name": "Beijing", "population": 100}'
 SELECT id FROM city WHERE population >= 100;
 ```
 
+# MySQL 兼容性
+
+- 修复了 MySQL 使用二进制类型数据创建 JSON 时错误的将其标记为 STRING TYPE 的 bug：
```
Comment (YangKeao) [inline data-type-json.md:30]: 还有一堆别的类型，比如 DATE，DATETIME，TIMESTAMP，在修了 :crying_cat_face:

File: data-type-json.md [hunk 2]
```diff
@@ -29,4 +25,85 @@ INSERT INTO city (id,detail) VALUES (1, '{"name": "Beijing", "population": 100}'
...
+create table t(e enum('a'));
+insert into t values ('a');
+mysql> select cast(e as json) from t;
+ERROR 3140 (22032): Invalid JSON text: The document root must not be followed by other values.
+```
+详情可见此 [issue](https://github.com/pingcap/tidb/issues/9999)
+
+- 支持对 JSON ARRAY/OBJECT 的 `order by`
+
+如果在 MySQL 中对 JSON ARRAY/OBJECT 使用 order by 会得到一个 warning，且结果与比较运算不一致:
```
Comment (qiancai) [inline data-type-json.md:71]:
```suggestion
当使用 `ORDER BY` 对 JSON Array 或 JSON Object 进行排序时，MySQL 会返回一个警告，且排序结果与比较运算结果不一致:
```

File: data-type-json.md [hunk 3]
```diff
@@ -29,4 +25,85 @@ INSERT INTO city (id,detail) VALUES (1, '{"name": "Beijing", "population": 100}'
...
+
+mysql> select j from t where j < json_array(5);
++--------------+
+| j            |
++--------------+
+| [1, 2, 3, 4] |
++--------------+
+1 row in set (0.00 sec)
+
+-- 在 MySQL 中下面 SQL 会抛出 warning: This version of MySQL doesn't yet support 'sorting of non-scalar JSON values'. 且结果与 `<` 比较结果不一致
```
Comment (qiancai) [inline data-type-json.md:88]:
```suggestion
-- 在 MySQL 中，执行以下 SQL 语句会返回警告 “This version of MySQL doesn't yet support 'sorting of non-scalar JSON values'. ”，且排序结果与 `<` 比较结果不一致。
```

File: data-type-json.md [hunk 4]
```diff
@@ -29,4 +25,85 @@ INSERT INTO city (id,detail) VALUES (1, '{"name": "Beijing", "population": 100}'
...
+| j            |
++--------------+
+| [1, 2, 3, 4] |
+| [5]          |
++--------------+
+2 rows in set (0.00 sec)
+```
+详情可见此 [issue](https://github.com/pingcap/tidb/issues/37506)
+
+- 暂不支持 JSON PATH 中范围选取的语法，以下 SQL 会在 TiDB 中报错:
```
Comment (qiancai) [inline data-type-json.md:100]:
```suggestion
- TiDB 暂不支持 JSON PATH 中范围选取的语法，以下 SQL 语句会在 TiDB 中报错:
```

File: data-type-json.md [hunk 5]
```diff
@@ -29,4 +25,85 @@ INSERT INTO city (id,detail) VALUES (1, '{"name": "Beijing", "population": 100}'
 SELECT id FROM city WHERE population >= 100;
 ```
 
+# MySQL 兼容性
+
+- 修复了 MySQL 使用二进制类型数据创建 JSON 时错误的将其标记为 STRING TYPE 的 bug：
+
+{{< copyable "sql" >}}
```
Comment (qiancai) [inline data-type-json.md:32]: “{{< copyable "sql" >}}“ 不再需要了，文档站会自动渲染增加复制 icon。麻烦请删除。

File: data-type-json.md [hunk 6]
```diff
@@ -29,4 +25,85 @@ INSERT INTO city (id,detail) VALUES (1, '{"name": "Beijing", "population": 100}'
...
+
+```sql
+create table test (a json);
```
Comment (qiancai) [inline data-type-json.md:35]: 请将本文中的所有 SQL 关键字改成大写的。

File: data-type-json.md [hunk 7]
```diff
@@ -29,4 +25,85 @@ INSERT INTO city (id,detail) VALUES (1, '{"name": "Beijing", "population": 100}'
...
+
+```sql
+create table t(e enum('a'));
+insert into t values ('a');
+mysql> select cast(e as json) from t;
+ERROR 3140 (22032): Invalid JSON text: The document root must not be followed by other values.
+```
+详情可见此 [issue](https://github.com/pingcap/tidb/issues/9999)
+
+- 支持对 JSON ARRAY/OBJECT 的 `order by`
```
Comment (qiancai) [inline data-type-json.md:69]:
```suggestion
- TiDB 支持使用 `ORDER BY` 对 JSON Array 或 JSON Object 进行排序。
```

File: data-type-json.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (qiancai) [reply data-type-json.md:30]: 如果这个 bug 不是我们修复的。TiDB 本身就没有这个问题，建议这里换个说法👆。

File: data-type-json.md [hunk 8]
```diff
@@ -29,4 +25,81 @@ INSERT INTO city (id,detail) VALUES (1, '{"name": "Beijing", "population": 100}'
...
+    -- TiDB 结果如下所示, 在 MySQL 中下面 SQL 结果为 `0, 1`。
+    mysql> SELECT JSON_EXTRACT(JSON_OBJECT('a', b'01010101'), '$.a') = "base64:type15:VQ==" AS r1, JSON_EXTRACT(a, '$.a') = "base64:type15:VQ==" AS r2 FROM test;
+    +------+------+
+    | r1   | r2   |
+    +------+------+
+    |    0 |    0 |
+    +------+------+
+    1 row in set (0.01 sec)
+    ```
+    详情可见此 [issue](https://github.com/pingcap/tidb/issues/37443)
```
Comment (qiancai) [inline data-type-json.md:56]:
```suggestion
    详情可见此 [issue](https://github.com/pingcap/tidb/issues/37443)。
```

File: data-type-json.md [hunk 9]
```diff
@@ -29,4 +25,81 @@ INSERT INTO city (id,detail) VALUES (1, '{"name": "Beijing", "population": 100}'
...
+
+- 当 `enum/set` 转换为 JSON 时，TiDB 会检查其格式正确性。下面 SQL 在 TiDB 中会报错:
```
Comment (qiancai) [inline data-type-json.md:58]:
```suggestion
- 当 `enum/set` 转换为 JSON 时，TiDB 会检查其格式正确性。例如，当执行下面的 SQL 语句时，TiDB 中会报错：
```

File: data-type-json.md [hunk 10]
```diff
@@ -29,4 +25,81 @@ INSERT INTO city (id,detail) VALUES (1, '{"name": "Beijing", "population": 100}'
...
+
+    ```sql
+    CREATE TABLE t(e ENUM('a'));
+    INSERT INTO t VALUES ('a');
+    mysql> SELECT CAST(e AS JSON) FROM t;
+    ERROR 3140 (22032): Invalid JSON text: The document root must not be followed by other values.
+    ```
+    详情可见此 [issue](https://github.com/pingcap/tidb/issues/9999)
```
Comment (qiancai) [inline data-type-json.md:66]:
```suggestion
    详情可见此 [issue](https://github.com/pingcap/tidb/issues/9999)。
```

File: data-type-json.md [hunk 11]
```diff
@@ -29,4 +25,81 @@ INSERT INTO city (id,detail) VALUES (1, '{"name": "Beijing", "population": 100}'
...
+
+# MySQL 兼容性
+
+- 当使用二进制类型数据创建 JSON 时，目前 MySQL 会将其误标记为 STRING 类型。
+
+    ```sql
+    CREATE TABLE test(a json);
+    INSERT INTO test SELECT json_objectagg('a', b'01010101');
+
+    -- TiDB 结果如下所示, 在 MySQL 中下面 SQL 结果为 `0, 1`。
```
Comment (qiancai) [inline data-type-json.md:47]:
```suggestion
    -- 在 TiDB 中，执行以下 SQL 语句返回结果如下所示。在 MySQL 中，执行以下 SQL 语句的结果为 `0, 1`。
```

File: data-type-json.md [hunk 12]
```diff
@@ -29,4 +25,81 @@ INSERT INTO city (id,detail) VALUES (1, '{"name": "Beijing", "population": 100}'
...
+
+    mysql> SELECT j FROM t WHERE j < JSON_ARRAY(5);
+    +--------------+
+    | j            |
+    +--------------+
+    | [1, 2, 3, 4] |
+    +--------------+
+    1 row in set (0.00 sec)
+
+    -- 在 MySQL 中，执行以下 SQL 语句会返回警告 “This version of MySQL doesn't yet support 'sorting of non-scalar JSON values'. ”，且排序结果与 `<` 比较结果不一致。
```
Comment (qiancai) [inline data-type-json.md:85]:
```suggestion
    -- 在 TiDB 中，执行以下 SQL 语句返回结果如下所示。在 MySQL 中，执行以下 SQL 语句会返回警告 “This version of MySQL doesn't yet support 'sorting of non-scalar JSON values'. ”，且排序结果与 `<` 比较结果不一致。
```

File: data-type-json.md [hunk 13]
```diff
@@ -29,4 +25,81 @@ INSERT INTO city (id,detail) VALUES (1, '{"name": "Beijing", "population": 100}'
...
+    mysql> SELECT j FROM t ORDER BY j;
+    +--------------+
+    | j            |
+    +--------------+
+    | [1, 2, 3, 4] |
+    | [5]          |
+    +--------------+
+    2 rows in set (0.00 sec)
+    ```
+    详情可见此 [issue](https://github.com/pingcap/tidb/issues/37506)
```
Comment (qiancai) [inline data-type-json.md:95]:
```suggestion
    详情可见此 [issue](https://github.com/pingcap/tidb/issues/37506)。
```

File: data-type-json.md [hunk 14]
```diff
@@ -29,4 +25,81 @@ INSERT INTO city (id,detail) VALUES (1, '{"name": "Beijing", "population": 100}'
...
+
+- 在 INSERT JSON 列时，TiDB 会将值隐式转换为 JSON
+
+    ```sql
+    CREATE TABLE t(col JSON);
+    -- TiDB INSERT 成功，MySQL 会报错: Invalid JSON text
```
Comment (qiancai) [inline data-type-json.md:101]:
```suggestion

    -- 在 TiDB 中，执行以下 INSERT 语句成功。在 MySQL 中，执行以下 INSERT 语句将返回 Invalid JSON text 错误。
```

File: data-type-json.md [hunk 15]
```diff
@@ -29,4 +25,81 @@ INSERT INTO city (id,detail) VALUES (1, '{"name": "Beijing", "population": 100}'
...
+# 使用限制
+
+- TiDB 暂不支持 JSON PATH 中范围选取的语法，以下 SQL 语句会在 TiDB 中报错:
+
+    ```sql
+    SELECT j->'$[1 to 2]' FROM t;
+    SELECT j->'$[last]' FROM t;
+    ```
+
+- 请勿使用 BR 工具将备份的 `>=6.3` 版本包含 JSON 列的数据恢复至 `<6.3` 版本的集群中，否则有几率出现无法查询 JSON 的问题。
```
Comment (qiancai) [inline data-type-json.md:37]:
```suggestion
- TiDB Backup & Restore（BR）在 v6.3.0 之前不支持恢复包含 JSON 列的数据。另外，任何版本的 BR 都不支持恢复包含 JSON 列的数据到 v6.3.0 之前的 TiDB 集群。
```

Comment (qiancai) [reply data-type-json.md:37]: 麻烦参考 https://docs.pingcap.com/zh/tidb/stable/character-set-gbk#%E7%BB%84%E4%BB%B6%E5%85%BC%E5%AE%B9%E6%80%A7 看下是否要这么改

File: data-type-json.md [hunk 16]
```diff
@@ -29,4 +25,81 @@ INSERT INTO city (id,detail) VALUES (1, '{"name": "Beijing", "population": 100}'
...
+    CREATE TABLE t(e ENUM('a'));
+    INSERT INTO t VALUES ('a');
+    mysql> SELECT CAST(e AS JSON) FROM t;
+    ERROR 3140 (22032): Invalid JSON text: The document root must not be followed by other values.
+    ```
+    详情可见此 [issue](https://github.com/pingcap/tidb/issues/9999)
+
+- TiDB 支持使用 `ORDER BY` 对 JSON Array 或 JSON Object 进行排序。
+
+    当使用 `ORDER BY` 对 JSON Array 或 JSON Object 进行排序时，MySQL 会返回一个警告，且排序结果与比较运算结果不一致:
```
Comment (qiancai) [inline data-type-json.md:70]:
```suggestion
    当使用 `ORDER BY` 对 JSON Array 或 JSON Object 进行排序时，MySQL 会返回一个警告，且排序结果与比较运算结果不一致：
```

**Comment by YangKeao** [comment]
I found a section in the aggregate documents: https://docs.pingcap.com/tidb/dev/aggregate-group-by-functions#unsupported-aggregate-functions.

image

Maybe we should delete these lines?

**Review by YangKeao — APPROVED** [review]
LGTM

File: data-type-json.md [hunk 17]
```diff
@@ -29,4 +25,87 @@ INSERT INTO city (id,detail) VALUES (1, '{"name": "Beijing", "population": 100}'
...
+- 暂不支持将 JSON 函数下推至 TiFlash。
+- TiDB 暂不支持 JSON PATH 中范围选取的语法，以下 SQL 语句会在 TiDB 中报错:
+
+    ```sql
+    SELECT j->'$[1 to 2]' FROM t;
+    SELECT j->'$[last]' FROM t;
+    ```
+
+- TiDB Backup & Restore（BR）在 v6.3.0 之前不支持恢复包含 JSON 列的数据。另外，任何版本的 BR 都不支持恢复包含 JSON 列的数据到 v6.3.0 之前的 TiDB 集群。
+- 请勿使用任何同步工具同步非标准 JSON 类型（例如 DATE/DATETIME/TIME 等）的数据。
```
[the remark this answers is not in the document]
Comment (xiongjiwei) [reply data-type-json.md:39]: 数据可以同步，但是类型会丢失，因为是 JSON 标准中没有这些类型。

File: data-type-json.md [hunk 18]
```diff
@@ -29,4 +25,86 @@ INSERT INTO city (id,detail) VALUES (1, '{"name": "Beijing", "population": 100}'
...
+    -- 在 TiDB 中，执行以下 SQL 语句返回结果如下所示。在 MySQL 中，执行以下 SQL 语句会返回警告 “This version of MySQL doesn't yet support 'sorting of non-scalar JSON values'. ”，且排序结果与 `<` 比较结果不一致。
+    mysql> SELECT j FROM t ORDER BY j;
+    +--------------+
+    | j            |
+    +--------------+
+    | [1, 2, 3, 4] |
+    | [5]          |
+    +--------------+
+    2 rows in set (0.00 sec)
+    ```
```
Comment (xiongjiwei) [inline data-type-json.md:98]:
```suggestion
    ```

```

File: data-type-json.md [hunk 19]
```diff
@@ -29,4 +25,87 @@ INSERT INTO city (id,detail) VALUES (1, '{"name": "Beijing", "population": 100}'
...
+- 暂不支持将 JSON 函数下推至 TiFlash。
+- TiDB 暂不支持 JSON PATH 中范围选取的语法，以下 SQL 语句会在 TiDB 中报错:
+
+    ```sql
+    SELECT j->'$[1 to 2]' FROM t;
+    SELECT j->'$[last]' FROM t;
+    ```
+
+- TiDB Backup & Restore（BR）在 v6.3.0 之前不支持恢复包含 JSON 列的数据。另外，任何版本的 BR 都不支持恢复包含 JSON 列的数据到 v6.3.0 之前的 TiDB 集群。
+- 请勿使用任何同步工具同步非标准 JSON 类型（例如 DATE/DATETIME/TIME 等）的数据。
```
Comment (qiancai) [inline data-type-json.md:39]:
```suggestion
- 请勿使用任何同步工具同步非标准 JSON 类型（例如 DATE、DATETIME、TIME 等）的数据。
```

File: data-type-json.md [hunk 20]
```diff
@@ -29,4 +25,87 @@ INSERT INTO city (id,detail) VALUES (1, '{"name": "Beijing", "population": 100}'
...
+    +--------------+
+    | [1, 2, 3, 4] |
+    | [5]          |
+    +--------------+
+    2 rows in set (0.00 sec)
+    ```
+
+    详情可见此 [issue](https://github.com/pingcap/tidb/issues/37506)。
+
+- 在 INSERT JSON 列时，TiDB 会将值隐式转换为 JSON
```
Comment (hfxsd) [inline data-type-json.md:102]:
```suggestion
- 在 INSERT JSON 列时，TiDB 会将值隐式转换为 JSON：
```

File: experimental-features.md [hunk 21]
```diff
@@ -34,7 +34,6 @@ aliases: ['/docs-cn/dev/experimental-features-4.0/','/zh/tidb/dev/experimental-f
 + 表达式索引 (Expression Index) 功能。表达式索引也叫函数索引，在创建索引时索引的字段不一定是一个具体的列，也可以是由一个或者多个列计算出来的表达式。对于快速访问那些基于计算结果的表非常有用。详情参阅：[表达式索引](/sql-statements/sql-statement-create-index.md)。（v4.0 实验特性）
 + [生成列](/generated-columns.md#生成列)。（v2.1 实验特性）
 + [自定义变量](/user-defined-variables.md#用户自定义变量)。（v2.1 实验特性）
-+ [JSON 数据类型](/data-type-json.md)及 [JSON 函数](/functions-and-operators/json-functions.md)。（v2.1 实验特性）
```
Comment (hfxsd) [inline experimental-features.md:37]: 这里的 （v2.1 实验特性） 要删掉吗？

**Review by hfxsd — APPROVED** [review]
Rest LGTM

File: sql-statements/sql-statement-create-index.md [hunk 22]
```diff
@@ -197,11 +197,14 @@ DROP INDEX idx1 ON t1;
...
-> +--------------------------------------------+
-> | @@tidb_allow_function_for_expression_index |
-> +--------------------------------------------+
-> | lower, md5, reverse, upper, vitess_hash    |
-> +--------------------------------------------+
+> +------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+> | @@tidb_allow_function_for_expression_index
+> 
+> 
+>                    |
```
Comment (hfxsd) [inline sql-statements/sql-statement-create-index.md:204]: 这行好像应该删掉？预览效果不太正常。

File: data-type-json.md [hunk 23]
```diff
@@ -29,4 +25,87 @@ INSERT INTO city (id,detail) VALUES (1, '{"name": "Beijing", "population": 100}'
...
+    SELECT j->'$[1 to 2]' FROM t;
+    SELECT j->'$[last]' FROM t;
+    ```
+
+- TiDB Backup & Restore（BR）在 v6.3.0 之前不支持恢复包含 JSON 列的数据。另外，任何版本的 BR 都不支持恢复包含 JSON 列的数据到 v6.3.0 之前的 TiDB 集群。
+- 请勿使用任何同步工具同步非标准 JSON 类型（例如 DATE、DATETIME、TIME 等）的数据。
+
+## MySQL 兼容性
+
+- 当使用二进制类型数据创建 JSON 时，目前 MySQL 会将其误标记为 STRING 类型。
```
Comment (songrijie) [inline data-type-json.md:43]: 这里是不是加一句，"而 TiDB 会维持正确的二进制类型"。

**Comment by songrijie** [comment]
rest LGTM

File: sql-statements/sql-statement-create-index.md (hunk 22 above)
Comment (xiongjiwei) [reply sql-statements/sql-statement-create-index.md:204]: 这里我觉得换一种方式吧，直接把他们都列出来，不用 SQL 结果来展示了，列表太长了

File: data-type-json.md [hunk 24]
```diff
@@ -29,4 +25,87 @@ INSERT INTO city (id,detail) VALUES (1, '{"name": "Beijing", "population": 100}'
...
+    SELECT j->'$[1 to 2]' FROM t;
+    SELECT j->'$[last]' FROM t;
+    ```
+
+- TiDB Backup & Restore（BR）在 v6.3.0 之前不支持恢复包含 JSON 列的数据。另外，任何版本的 BR 都不支持恢复包含 JSON 列的数据到 v6.3.0 之前的 TiDB 集群。
+- 请勿使用任何同步工具同步非标准 JSON 类型（例如 DATE、DATETIME、TIME 等）的数据。
+
+## MySQL 兼容性
+
+- 当使用二进制类型数据创建 JSON 时，目前 MySQL 会将其误标记为 STRING 类型，而 TiDB 会维持正确的二进制类型。
```
Comment (xiongjiwei) [inline data-type-json.md:43]:
```suggestion
- 当使用二进制类型数据创建 JSON 时，目前 MySQL 会将其误标记为 STRING 类型，而 TiDB 会保持正确的二进制类型。
```

**Pull request merged**